### DISCUSSION TITLE
Smaller margins in dual page mode?

### DISCUSSION THREAD

**Discussion by WS64** [koreader/koreader#7177] [Q&A]
Is it possible to make the borders on the L/R margins in the middle smaller?
Here in dual page portrait mode it is extrem, but in landscape mode it is not much different.
I have set L/R margins to zero already, but there is still a lot of space in the middle wasted.
![Reader_2021-01-19_194527](https://user-images.githubusercontent.com/7792441/105210316-2e284d00-5b4b-11eb-9041-362dfe4189e7.png)

**Comment by poire-z** [comment]
It is hardcoded at https://github.com/koreader/crengine/blob/c4b6b9ee2a5e123f20955b80b29b3a0063863844/crengine/src/lvdocview.cpp#L2579-L2611 to be at min 1.2em (= 1.2 your normal font size) - unless I messed the computation.
So, it decreases as your font size decreases. May be your text itself has a font-size: 0.8 and is smaller than the normal (1rem) font size, so this large gap ? You can check that by enabling the "Ignore publisher font sizes" style tweak.

For me, with the text using the normal font size, it feels alright at various font sizes:

![image](https://user-images.githubusercontent.com/24273478/105427971-a592d500-5c4e-11eb-9780-0e3c5475204f.png)

![image](https://user-images.githubusercontent.com/24273478/105427982-adeb1000-5c4e-11eb-931f-2b377797651b.png)

![image](https://user-images.githubusercontent.com/24273478/105427996-b80d0e80-5c4e-11eb-8ee0-dfb20fbab78a.png)

![image](https://user-images.githubusercontent.com/24273478/105428038-ca874800-5c4e-11eb-9373-3ea26e739216.png)

We don't want to go as low as the L or R margin, we need more between the 2 columns for readability.

**Reply by NiLuJe** [reply]
I don't have anything against the current behavior, so, playing devil's advocate here:

What about using the value of L+R margins instead?
(... which, granted, just makes one itch to just make that configurable and be done with it ;p).

**Reply by poire-z** [reply]
Because 0+0 = 🤯 ... and the 2 columns stuck to each others.
(And 2cm + 2cm = 4 cm middle margins is too much - see my comments in the code linked above :) I didn't think that too much, but the results looked alright...)
(And a configurable would have you adjust it each time you change font size - the minimal needed margin for readability feels related to the font size, to me.)

**Reply by WS64** [reply]
I guess this file is not available in the koreader folder? I'd like to patch it since this min_middle_margin is too big...
What does 0+0=smiley mean?
Yes, if you would set the LR margins to zero they would (and should) stuck together.
But even right now you can't go down to zero by the graphical interface but have to enter the numbers manually, so normally you would not even reach zero.

**Reply by poire-z** [reply]
No, it's some code that is compiled and make it into libcrengine.so - so you can't tweak it.

0+0 = head exploding because no space between columns.

But please try what I suggested:
>  May be your text itself has a font-size: 0.8 and is smaller than the normal (1rem) font size, so this large gap ? You can check that by enabling the "Ignore publisher font sizes" style tweak.

**Reply by WS64** [reply]
No difference for the margins...
![Reader_2021-01-22_192647](https://user-images.githubusercontent.com/7792441/105530395-06222080-5ce8-11eb-84e6-38e021015d7a.png)
![Reader_2021-01-22_192718](https://user-images.githubusercontent.com/7792441/105530396-07534d80-5ce8-11eb-9e8c-162c4e954260.png)

**Reply by poire-z** [reply]
I don't undestand why you get such a huge middle margin.
2 questions:
- on my screenshots above, does the margin seems good (small enough) to you, for the various font sizes? And do you notice the different vs your screenshots (I do).
- on your book: try to disable Embedded Styles in the bottom right menu (in case your book have some additional margins/paddings...)
Sorry, I can't draw straight, but you should get the same as in the top header:
![image](https://user-images.githubusercontent.com/24273478/105533441-6fa42e00-5cec-11eb-98e1-9413b03b52f3.png)

**Reply by WS64** [reply]
Here's without embedded styles. 
I cannot see a difference in the margins, only top and bottom, but that is because of widows and orphans in embedded styles.
![Reader_2021-01-22_203314](https://user-images.githubusercontent.com/7792441/105538373-974ac480-5cf3-11eb-9e39-1549edcb75d4.png)
Yours are a bit better, but still too big.

**Reply by WS64** [reply]
Just for the fun of it, I edited the history file and change this to some settings not available in the interface:
    ["copt_h_page_margins"] = {
        [2] = -6,
        [1] = -6,
    },
From -9 on or so the text is not completely visible anymore.
A tiny bit better, now the text is at least on the same level as the top bar.
But no change to the margin.

![Reader_2021-01-22_211844](https://user-images.githubusercontent.com/7792441/105541154-a2075880-5cf7-11eb-8456-7cb26889f6ac.png)

**Reply by poire-z** [reply]
Mhhh, there's still this big **A** - so may be it's `<big>` - or this A is in the regular font size and all the rest of the text is `<small>`. Can you try with top menu 2nd item > `Style > none`.
Can you also try with another font (without bold if you have it enabled)?

I would like you to look at the code I linked above, understand and check the logic, and possibly compile yourself a crengine with another value than `1.2` - or some other logic change that you feel would make it ok - and actually see if (and down to which point) it makes readability harder.
If it's just this 1.2 that limits you from getting a smaller margin, may be this can be tweaked via a crengine cr3.ini setting (that we may or may not make available in the frontend menu).

(I don't read in dual page mode, but I just tried it with various margin and font sizes - and I hardly feel the need to have it smaller. Yes, it surely looks larger than the left and right margins, but it's necessary to keep your eyes from crossin columns :) but I don't get the large one you get on your screenshot.)

**Reply by poire-z** [reply]
> Sorry, I can't draw straight, but you should get the same as in the top header:

I might have been wrong with that: I noticed a difference on my Kobo - so the top header bar is not using the middle margin we compute in that code snippet.

**Reply by WS64** [reply]
I don't think I will be able to compile crengine unless there is a way to do it in Windows.

The big A:
```
<p class="new"><span class="firstletter">»A</span>fter him!« urged young Squint as I fingered the throttle irresolutely.</p>
with 
.new {text-indent:0;}
.firstletter {font-size:175%;line-height:1em;}
```

I think it should look more like this (Amazing Stories, 1952-09):
![Bild1](https://user-images.githubusercontent.com/7792441/105545963-04168c80-5cfd-11eb-886d-3a86db7a93b9.jpg)

**Reply by WS64** [reply]
LR back to 0, style: none:
![Reader_2021-01-22_220221](https://user-images.githubusercontent.com/7792441/105546426-aa629200-5cfd-11eb-9208-552074316e42.png)

**Reply by poire-z** [reply]
`Style: none` + `Embedded styles off` please :)

Can you try with a Wikipedia Save as EPUB epub, with another font we know like Noto Serif?

**Reply by WS64** [reply]
I don't see any option to export Wikiepedia articles to Epub, only PDF.
Anyway,here's my book again, style none and embedded styles off, with my original font and with Noto.
![Reader_2021-01-22_224259](https://user-images.githubusercontent.com/7792441/105571554-16bab100-5d51-11eb-90ae-65fe85d206a2.png)
![Reader_2021-01-22_224336](https://user-images.githubusercontent.com/7792441/105571555-17534780-5d51-11eb-8772-c3b932aa81eb.png)

**Reply by Frenzie** [reply]
@WS64 In KOReader itself, if you search for a Wikipedia article you can choose to save it as EPUB.

**Reply by WS64** [reply]
Ahhh, ok, but I can't: Saving Wikipedia article failed or interrupted
I guess this is because it tries to save the article on my read-only mounted SD card instead of the Koreader folder? JUst a guess. Anyway, I can't save it!
Crash log:
PING 192.168.0.1 (192.168.0.1): 56 data bytes
64 bytes from 192.168.0.1: seq=0 ttl=64 time=30.283 ms

--- 192.168.0.1 ping statistics ---
1 packets transmitted, 1 packets received, 0% packet loss
round-trip min/avg/max = 30.283/30.283/30.283 ms
ffi.load: z
01/23/21-08:42:36 WARN  Wikipedia.createEpub pcall: true false

**Reply by Frenzie** [reply]
I think that by default it will try to save it in the folder you've set as HOME. If that's a read-only location, then yeah, that wouldn't work. :-) You can select alternative locations in the settings (i.e., in the current folder next to the open document, in ~/Wikipedia, or in any folder of your choice).

**Reply by WS64** [reply]
got it!
![Reader_2021-01-23_094708](https://user-images.githubusercontent.com/7792441/105575119-6d33e980-5d69-11eb-986a-16067e55c68f.png)

**Reply by poire-z** [reply]
On this one, please enable styles (so text is justified and we can see the real page widths and so the real margins :) I know that in Wikipedia EPUBs, we don't add any margins or tweaks, unlike what may happens in books.

(I also fear you might have some other CSS or Lua tweaks that may be at play with that, do you?)

**Reply by WS64** [reply]
![Reader_2021-01-23_124737](https://user-images.githubusercontent.com/7792441/105577359-5d240600-5d79-11eb-9ee8-667264651676.png)

And same with my original font:
![Reader_2021-01-23_125008](https://user-images.githubusercontent.com/7792441/105577391-9eb4b100-5d79-11eb-9ea6-1d1e2354cb11.png)

And no other tweaks!

**Reply by poire-z** [reply]
That last screenshot feels a lot better than your previous ones - and how I do expect the middle margin to be sized !
At least, it matches what I expect. Do you also see the differences, any reason for it? I'd like to know why you got it larger with your previous book - because if there are other stuff involved, we might change focus on what needs to be done.

**Reply by WS64** [reply]
Agreed. I found the culprit, the css for this book included
body {margin:8pt;}
I thought I could simply get rid of this by putting a css file in the styletweaks folder with the line
body {margin:0pt;}
and of couse enabled it in the menu.
But it did not help.
Then I edited the book and changed the line there, works.
Do I misunderstand how the styletweaks works?

**Reply by poire-z** [reply]
> I found the culprit, the css for this book included

These should have been disabled when you toggle Embedded Styles to off - but we didn't see any effect.

> I thought I could simply get rid of this by putting a css file in the styletweaks folder with the line
body {margin:0pt;}

Our style tweaks are applied before the book CSS. So you need body {margin:0pt **!important**;}.
But we already have that one `Pages > Ignore publisher page margins`.

> Agreed.

So, are you now somehow fine with the sizing of the middle margin? And/or how it reacts to L/R margins and font size changes?

**Reply by WS64** [reply]
ok, !important did the trick.
It is for sure better that way. But to be honest, when I put the margins to zero I'd expect "zero", so the columns should stick together.
If there is a chance to make the 1.2em configurable I would prefer that.
Epecially in portrait mode I still think there is too much space wasted in the middle!
![Reader_2021-01-23_141202](https://user-images.githubusercontent.com/7792441/105579153-058b9780-5d85-11eb-8cc9-ad2ba4d60b9d.png)

**Reply by poire-z** [reply]
> But to be honest, when I put the margins to zero I'd expect "zero", so the columns should stick together.

Mhhh, you put the L/R margins to 0, and you would expect KOReader to not care much and put 0 in the middle and make the book unreadable ?! I'm sorry to hear hat you expect so little of us :)

@zwim: any thought of the middle margin size, after a few days of playing with dual page in portrait mode?

**I f*cking hate this github Discussion stuff, it has no visibility, no linkability to the issue where we're currently discussing dual pages... I always have to click "view more" to read the latest posts... Please please please give up on it and use regular Issues...**

**Reply by WS64** [reply]
I agree on the gitthub discussion, really hard to keep track.
It was suggested I use it, and it was worth a try, but the "issue" section really is better!

**Reply by hius07** [reply]
> any thought of the middle margin size

Who told about two items only in the bottom menu and a room for more settings? :)

**Reply by Frenzie** [reply]
@poire-z 
> I always have to click "view more" to read the latest posts...

I really hate that about GH PRs btw. I haven't yet noticed it in these discussion things.

I do like the theory behind them, namely that your issues are things more directly relating to the code and these are… discussions about any old thing. I'm fairly neutral on them I think but I'm inclined to agree it's more of a poor forum and not really a part of GH with added value.

**Comment by NiLuJe** [comment]
@poire-z: Wild idea: what about picking up the "line spacing" prop instead of hard-coding it to 1.2em?

**Reply by poire-z** [reply]
I would think these are quire orthogonal - but I dunno, may be. @WS64 seems to use smaller line spacing that what I'm used to. is that right ? Or is it just your font that have a small line spacing?

Btw, I might be willing to reduce that 1.2em it it's fine for most real dual page readers :) I picked that value quite randomly. I'm also seeing I might apply it on both sides of the middile margin, so it might be 2.4em ?! (I couldn't really get back into that bit of code I linked above :)

**Reply by NiLuJe** [reply]
AFAICT, what it does is essentially use 1.2em instead of the greater of L OR R if L + R < 1.2em).
I think the logic is fairly convoluted because margins can be negative, or something related to the initial comment, hence the L+R subtraction?

Which I guess, for most use-cases, could be simplified to something like
```C++
// We want at least 1.2em
int floor_margin = 1.2 * m_font_size;
// We prefer using the greater of L or R first, though
int target_margin = m_pageMargins.left > m_pageMargins.right ? m_pageMargins.left : m_pageMargins.right;
// The amount of space that will be baked in the rects later
int baked_margin = m_pageMargins.right + m_pageMargins.left;
// Which gives us the final extra margin we actually need to add (or subtract) given that.
int margin = baked_margin - MAX(target_margin, floor_margin);
```

(I'm assuming there's a MIN/MAX macro somewhere in the CRe codebase ^^).

This should ensure that negative page margins jump straight to 1.2em instead of... whatever's happening right now ^^.

(AFAICT, the actual application to the rects is then accurate, as that's the *entire* middle margin, hence the / 2 to split it between the two cols).

**Reply by Frenzie** [reply]
Vaguely related, CSS has a column-gap property. Perhaps it could be of interest to repurpose this code as CSS columns and to add support for that property to customize, but perhaps the spec for columns isn't really compatible with what we're doing here. (Or even if it is, maybe we still want to treat it with a special -cre prefix or some such to avoid conflicts.)

**Reply by poire-z** [reply]
> because margins can be negative

Oh my :) if they can, I guess that's not voluntary :) and things can be "undefined behaviour" !

> Which I guess, for most use-cases, could be simplified to something like

Bit busy with other stuff to get back into that, so if you feel like baking it, feel free :)

> the spec for columns isn't really compatible with what we're doing here

Indeed, at this point, we're far away up from the document and its CSS : we're just defining crengine view/page geometry - in which is will render the document.
(And no support CSS column, that's the kind of things so rarely used and possibly complicated to implement.)

**Reply by Frenzie** [reply]
Rarely used? Boy, I've been using it constantly since 2012. :P

**Reply by Frenzie** [reply]
@poire-z Oh, I just saw that "view more" here. Okay, that's really annoying.

**Reply by NiLuJe** [reply]
> @poire-z Oh, I just saw that "view more" here. Okay, that's really annoying.

Huh, yeah, first time I've noticed it, too. I was fairly okay with this it showed up, actually ^^.

**Reply by poire-z** [reply]
> first time I've noticed it, too

That means you haven't noticed all the other ones :) There was one just here, there is another one above for the sub-discussion with screenshots - an earlier discussion may continue, and you won't notice it if you scroll to bottom like we're used to with issues/PRs...
The only real way to follow this is via the mail notifictions we get...

And hop, hitting <kbd>Reply</kbd> and having this post ending in a nearly invisible `View more`... :/

**Reply by NiLuJe** [reply]
@poire-z: Okay, after testing, rewriting the logic in a possibly-easier-to-grok yielded similar results...

And it actually feels decent... as long as there aren't margins/padding in the CSS...
Do we have a way to know about those at this point?

(Romeo & Juliet in the tests is a perfect example of that, it has ginormous margins).

```
# With L/R set to 18px, and the book's page margins disabled.
min_middle_margin=48
max_middle_margin=18
middle_margin=36
additional_middle_margin=12

# Which matches my commented logic ;p
floor_margin=48
target_margin=18
baked_margin=36
Increasing margin to floor_margin
margin=12
```

```C++
        // We want at least 1.2em
        int floor_margin = 1.2 * m_font_size;
        fprintf(stderr, "floor_margin=%d\n", floor_margin);
        // We prefer using the greater of L or R first, though
        int target_margin = m_pageMargins.left > m_pageMargins.right ? m_pageMargins.left : m_pageMargins.right;
        fprintf(stderr, "target_margin=%d\n", target_margin);
        // The amount of space that will be baked in the rects later
        int baked_margin = m_pageMargins.right + m_pageMargins.left;
        fprintf(stderr, "baked_margin=%d\n", baked_margin);

        // Which gives us the final extra margin we actually need to add (or subtract) given that.
        int margin = 0;
        // If our target margin is smaller than floor, use the floor
        if (target_margin < floor_margin) {
            // If the baked in marging is *also* smaller than the floor, make sure we add enough space to reach the floor
            if (baked_margin < floor_margin) {
                fprintf(stderr, "Increasing margin to floor_margin\n");
                margin = floor_margin - baked_margin;
            } else {
                fprintf(stderr, "Using floor_margin\n");
                margin = baked_margin - floor_margin;
            }
        } else {
            // Here, baked_margin cannot be smaller than target_margin, because it's essentially target_margin + the opposite margin.
            fprintf(stderr, "Using target_margin\n");
            margin = baked_margin - target_margin;
        }
        fprintf(stderr, "margin=%d\n", margin);
```

**Reply by NiLuJe** [reply]
With the crappy CSS:

![Screenshot_20210123_232731](https://user-images.githubusercontent.com/111974/105615651-b1a79f80-5dd2-11eb-8b1a-55ec4ccd273b.png)

Nerfed by the style tweak:
![Screenshot_20210123_232044](https://user-images.githubusercontent.com/111974/105615614-62616f00-5dd2-11eb-962c-066d9ec02851.png)

**Comment by poire-z** [comment]
It's lacking what you do with the "margin" at the end.
I had to rewrite "my" code with your naming to get to understand your new terminology:
```c
int floor_margin = 1.2 * m_font_size;
int target_margin = m_pageMargins.left > m_pageMargins.right ? m_pageMargins.left : m_pageMargins.right;
int margin = 0;
int baked_margin = m_pageMargins.right + m_pageMargins.left;
if (baked_margin < floor_margin) {
        margin = floor_margin - baked_margin;
}
else if (baked_margin > target_margin) {
        if (target_margin > floor_margin)
                margin = target_margin - baked_margin; // negative
        else
                margin = floor_margin - baked_margin; // negative
}
// Note: with negative values, we allow these 2 m_pageRects to
// overlap. But it seems there is no issue doing that.
m_pageRects[0].right = middle - margin / 2;
m_pageRects[1].left = middle + margin / 2;
```
And had to re-picture what are rects, and m_pageMargins is a false rect, just a holder of 4 values. And realized that "middle" is just a point.
I'm still very confused by both our codes :) The good thing is that we get the same result with different thinkings (are they really equivalent?), so, somehow we must both be right and that validates any algo :)

**Reply by NiLuJe** [reply]
Oh, yeah, the end seemed to make sense as-is ;p. i.e., you've got the middle of the screen, you've just computed how much of a gap, total, you want between columns, so, just stop one column at half of `margin` on the left of that and start the other at half of `margin` on the right ;).

The main "issue" I have with the logic that way around is that, technically, `baked_margin > target_margin` is *always* true, because baked_margin is target_margin + the opposite margin, so, even with super-weird margins, they're at worst equal ;p.

**Reply by NiLuJe** [reply]
What works decently well (but with generally tighter margins), is using L+R as the target gap, and the user's interline as the floor.

```C++
        fprintf(stderr, "interline=%d\n", m_def_interline_space);
        // We want at least an interline's worth of column gap
        int floor_gap = (m_def_interline_space / 100.f) * m_font_size;
        fprintf(stderr, "floor_gap=%d\n", floor_gap);
        // But we prefer using the the sum of the L+R margins as the column gap
        int target_gap = m_pageMargins.right + m_pageMargins.left;
        fprintf(stderr, "target_gap=%d\n", target_gap);
        // The page margins happen to *already* be baked into the pageRects later on,
        // so take that into account to compensate
        int baked_margin = m_pageMargins.right + m_pageMargins.left;
        fprintf(stderr, "baked_margin=%d\n", baked_margin);

        // Which gives us the final size of the column gap we'll actually need to add (or subtract) given that.
        int gap = 0;
        // If our target gap is smaller than floor, use the floor
        if (target_gap < floor_gap) {
            // If the baked in margin is *also* smaller than the floor gap, make sure we *add* enough space to reach the floor
            if (baked_margin < floor_gap) {
                fprintf(stderr, "Increasing margin to floor_gap\n");
                gap = floor_gap - baked_margin;
            } else {
                fprintf(stderr, "Using floor_gap\n");
                gap = baked_margin - floor_gap;
            }
        } else {
            // Here, baked_margin cannot be smaller than target_margin, they're always equal ;).
            fprintf(stderr, "Using target_gap\n");
            gap = baked_margin - target_gap;
        }
        fprintf(stderr, "gap=%d\n", gap);
        // NOTE: with negative values, we allow these 2 m_pageRects to
        // overlap, which appears to be harmless.
        // middle is at the middle of the screen, which becomes the middle of our column gap,
        // so, just set the left column's right edge at half a gap's worth on the left of that,
        // and the right column's left edge at half a gap's worth on the right of that.
        m_pageRects[0].right = middle - gap / 2;
        m_pageRects[1].left = middle + gap / 2;
```

Lit3rata, with a small native interline, and KO set to 110%
![Screenshot_20210124_181307](https://user-images.githubusercontent.com/111974/105637894-267ce700-5e70-11eb-96ac-33ba42bdd60e.png)

Noto Serif, with its giant native interline, and KO set to 80%
![Screenshot_20210124_181343](https://user-images.githubusercontent.com/111974/105637914-46140f80-5e70-11eb-85b4-a8f5baabd46d.png)

**Reply by NiLuJe** [reply]
e.g., if I were actually the target audience, that'd work for me. Keep in mind that I tend to favor very small margins and very small interlines, though, so, YMMV ;).

**Reply by poire-z** [reply]
Well, I'm not a fan of that kind of setting re-use/intervening :)
I guess I'll just switch the 1.2em to 0.8em (0.6em was really too small, 0.8em looked almost OK on the emulator, will have to check on my Kobo).

![image](https://user-images.githubusercontent.com/24273478/105638712-81184200-5e74-11eb-9cb2-1837b066dafc.png)

![image](https://user-images.githubusercontent.com/24273478/105638717-94c3a880-5e74-11eb-8973-6790a8202f39.png)

![image](https://user-images.githubusercontent.com/24273478/105638735-a7d67880-5e74-11eb-831a-90a52951fdf3.png)

People wanting a larger one could just increase L/R margins:

![image](https://user-images.githubusercontent.com/24273478/105638789-01d73e00-5e75-11eb-879f-9a767ea93ab9.png)

(so we can be done with this - and this hellish GitHub Discussion. :)

**Reply by NiLuJe** [reply]
That works, too ;).
